src/tilings.jl

```diff
@@ -106,100 +111,100 @@ t|   |r
 cycletetrahedrongroup(q::Int, r::Int, s::Int, t::Int) = tetrahedrongroup([q, 2, t, r, 2, s])
 
 """
-    normalsubgroups(g::ReflectionGroup, maxindex::Int)
+    normalsubgroups(G::ReflectionGroup, maxindex::Int)
 
-Return all normal subgroups of `g` with index up to `maxindex`.
+Return all normal subgroups of `G` with index up to `maxindex`.
 """
-function normalsubgroups(g::ReflectionGroup, maxindex::Int)
-    gr = GAP.Globals.LowIndexNormalSubgroupsSearchForAll(g.group, maxindex)
-    lns = GAP.Globals.List(gr)
-    sbgrps = Vector{GapObj}()
-    len = GAP.Globals.Length(lns)
-    for i = 1:len
-        push!(sbgrps, GAP.Globals.Grp(lns[i]))
-    end
+function normalsubgroups(G::ReflectionGroup, maxindex::Integer)
+    lins_search = GAP.Globals.LowIndexNormalSubgroupsSearchForAll(__gapobj(G), maxindex)
+    sbgrps = GapObj[H for H in GAP.Globals.List(lins_search)]
     return sbgrps
-end
+end 
 
 """
-    fixedpointfree(subgroup::GapObj, g::ReflectionGroup)
+    fixedpointfree(subgroup::GapObj, G::ReflectionGroup)
 
-Return `true` if the `subgroup` of `g` is fixed-point free; otherwise `false`.
+Return `true` if the `subgroup` of `G` is fixed-point free; otherwise `false`.
 """
-function fixedpointfree(subgroup::GapObj, g::ReflectionGroup)
-    hom = GAP.Globals.NaturalHomomorphismByNormalSubgroup(g.group, subgroup)
-    fpf = true
-    G = GapObj(())
-    for i in 1:g.dimension
-        fpf = fpf && GAP.Globals.Image(hom, g.generators[i]) != G
-    end
-    i = 1
-    for pair in combinations(1:g.dimension, 2)
-        for j in 1:g.orders[i] - 1
-            fpf = fpf && GAP.Globals.Image(hom, (g.generators[pair[1]] * g.generators[pair[2]])^j) != G
+function fixedpointfree(subgroup::GapObj, G::ReflectionGroup)
+    hom = GAP.Globals.NaturalHomomorphismByNormalSubgroup(__gapobj(G), subgroup)
+    h(g) = GAP.Globals.Image(hom, g)
+    fixedpoint = any(isone ∘ h, gens(G))
+    fixedpoint && return false
+    
+    for g1 in gens(G)
+        for (g2, ord) in zip(gens(G), orders(G))
+            fixedpoint = any(isone ∘ h, (g1*g2^j for j in 1:ord - 1))
```

Comment:
We need `(g1*g2)^j` here I think

Reply:
And I think the new implementation is different. We need to check the image of `(g1*g2)^j for j in 1:orders(1) - 1`, `(g1*g3)^j for j in 1:orders(2)-1`, ... 
How about this implementation?
```
for (comb, ord) in zip(combinations(gens(G), 2), orders(G))
    fixedpoint = any(isone ∘ h, (comb[1]*comb[2])^j for j in 1:ord - 1))
end
```

Reply:
yes, it was not clear to me what is the iteration space here;
for every pair `(g1, g2)` what are the possible orders?

Note: `zip(combinations(...), orders(G))` won't work, since zip has length of the shortest of the iterators

Reply:
Each item in `orders` is for a specific `g_i` and `g_j`. Specifically if we have `n` generators then `orders` should be length `n` choose 2, with the first item being the order for `g1` and `g2`, the second being the order for `g1` and `g3`, ... , the order for `g1` and `g_n`, the order for `g2` and `g3`, ..., the order for `g_{n-1}` and `g_n`. That's why I wanted to use `combinations()`.

Reply:
Ah, I see. This part of the code was rather opaque for me; It also depends on the order of elements in `combinations()`, which seems to me like an implementation detail.

----
The clean solution would be creating a thin matrix wrapper around `orders` vector:
```julia
struct Orders <: AbstractMatrix{Int}
    n::Int
    vec::Vector{Int} # or any other storage you find convenient
end
Base.size(o::Orders) = (o.n, o.n)

function Base.getindex(orders::Orders, i::Integer, j::Integer)
    # hide computation of linear index here
    idx = some_function_of(size(orders), i, j) # seems like a triangular indexing of sort
    return orders.vec[idx]
end
```

and then in `ReflectionGroup` storing just `orders::Vector{Integer}` store `orders::Orders` and turn this loop into
```julia
    for (i, g1) in enumerate(gens(G))
        for (j, g2) in enumerate(gens(G))
            k = orders(G)[i,j]
            fixedpoint = any(isone ∘ h, ((g1*g2)^j for j in 1:k - 1))
            [...]
```

@MikeVasmer can you point me to an example function where `orders` are created?
----
an immediate solution would be to do
```julia
    for ((g1, g2), ord) in zip(combinations(gens(G), 2), orders(G))
        fixedpoint = any(isone ∘ h, (g1*g2^j for j in 1:ord - 1))
        [...]
```
which relies on the particular order of `orders(G)` to be coupled with the order in which `combinations(..., 2)` are produced.

----
what do you think @MikeVasmer?

Reply:
Hello, sorry for my very slow response. I think the first example seems best as then, as you say, orders does not have to match combinations. At the moment, orders is just manually created in the `triangle_group` and `tetrahedron_group`, e.g. `return ReflectionGroup(g, [g.:1, g.:2, g.:3], [l, m, n], 3)`, where `[l, m, n]` are the orders. I am thinking that this implementation is also bad as the user is assumed to know how orders are indexed. I think your first solution is the best although maybe it's just easier to store the values as a symmetric matrix so e.g. for the `[l, m, n]` example above the matrix would be `[1 l m; l 1 n; m n 1]`. Then the user would just need to supply the entries above the diagonal.

Reply:
And then we could implement a generalized `trianglegroup` as follows:

```julia
function simplexgroup(orders::Orders)
    f = GAP.Globals.FreeGroup(orders.n)
    gens = GAP.Globals.GeneratorsOfGroup(f)
    relations = [gens[i]^2 for i in range(1, orders.n)]
    for (i, g1) in enumerate(gens)
        for (j, g2) in enumerate(gens)
            if j > i append!(relations, [(g1*g2)^orders[i,j]]) end
        end
    end
    g = f / GapObj(relations)
    return ReflectionGroup(g, gens, orders, orders.n)
end   
```

Reply:
We also don't need to check all the i and j in the fixedpoint calculation, maybe we should add a `if j > i`? (we can also exclude the i=j case here)

Reply:
Orders is really the Coxeter matrix https://en.wikipedia.org/wiki/Coxeter_group so I changed the name to reflect this